File: binding/go/cheetah_test.go
// Copyright 2022-2024 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is
// located in the "LICENSE" file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.
//

// Go binding for Cheetah Speech-to-Text engine.

package cheetah

import (
	"encoding/binary"
	"encoding/json"
	"flag"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

type LanguageTests struct {
	language                   string
	testAudioFile              string
	transcript                 string
	punctuations               []string
	errorRate                  float32
}

var (
	testAccessKey         string
	cheetah               Cheetah
	languageTests		  []LanguageTests
)

func TestMain(m *testing.M) {

	flag.StringVar(&testAccessKey, "access_key", "", "AccessKey for testing")
	flag.Parse()

	languageTests = loadTestData()
	os.Exit(m.Run())
}

func appendLanguage(s string, language string) string {
	if language == "en" {
		return s
	} else {
		return s + "_" + language
	}
}

func loadTestData() []LanguageTests {
	content, err := ioutil.ReadFile("../../resources/.test/test_data.json")
	if err != nil {
		log.Fatalf("Could not read test data json: %v", err)
	}

	var testData struct {
		Tests struct {
			LanguageTests []struct {
				Language                  string  `json:"language"`
				AudioFile                 string  `json:"audio_file"`
				Transcript                string  `json:"transcript"`
				Punctuations 			  []string  `json:"punctuations"`
				ErrorRate                 float32 `json:"error_rate"`
			} `json:"language_tests"`
		} `json:"tests"`
	}
	err = json.Unmarshal(content, &testData)
	if err != nil {
		log.Fatalf("Could not decode test data json: %v", err)
	}

	for _, x := range testData.Tests.LanguageTests {
		languageTestParameters := LanguageTests{
			language:                  x.Language,
			testAudioFile:             x.AudioFile,
			transcript:                x.Transcript,
			punctuations: 			   x.Punctuations,
			errorRate:                 x.ErrorRate,
		}

		languageTests = append(languageTests, languageTestParameters)
	}

	return languageTests
}

func TestVersion(t *testing.T) {
	cheetah = NewCheetah(testAccessKey)
	err := cheetah.Init()
	if err != nil {
		log.Fatalf("Failed to init cheetah with: %v", err)
	}

	defer func() {
		err := cheetah.Delete()
		if err != nil {
			log.Fatalf("Failed to release resources: %s", err)
		}
	}()

	if reflect.TypeOf(Version).Name() != "string" {
		t.Fatal("Unexpected version format.")
	}
	if Version == "" {
		t.Fatal("Failed to get version.")
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func levenshteinDistance(transcriptWords, referenceWords []string) int {
	m, n := len(transcriptWords), len(referenceWords)
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := 0; i <= m; i++ {
		dp[i][0] = i
	}
	for j := 0; j <= n; j++ {
		dp[0][j] = j
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			cost := 0
			if !strings.EqualFold(transcriptWords[i-1], referenceWords[j-1]) {
				cost = 1
			}
			dp[i][j] = min(dp[i-1][j]+1,
				min(dp[i][j-1]+1,
					dp[i-1][j-1]+cost))
		}
	}
	return dp[m][n]
}

func getWordErrorRate(transcript, reference string) float32 {
	transcriptWords := strings.Fields(transcript)
	referenceWords := strings.Fields(reference)

	dist := levenshteinDistance(transcriptWords, referenceWords)
	return float32(dist) / float32(len(referenceWords))
}

func runProcessTestCase(
	t *testing.T,
	language string,
	testAudioFile string,
	referenceTranscript string,
	punctuations []string,
	targetErrorRate float32,
	enableAutomaticPunctuation bool) {

	modelPath, _ := filepath.Abs(filepath.Join("../../lib/common", appendLanguage("cheetah_params", language)+".pv"))

	cheetah = NewCheetah(testAccessKey)
	cheetah.ModelPath = modelPath
	cheetah.EnableAutomaticPunctuation = enableAutomaticPunctuation
	err := cheetah.Init()
	if err != nil {
		log.Fatalf("Failed to init cheetah with: %v", err)
	}

	defer func() {
		err := cheetah.Delete()
		if err != nil {
			log.Fatalf("Failed to release resources: %s", err)
		}
	}()

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", testAudioFile))

	data, err := ioutil.ReadFile(testAudioPath)
	if err != nil {
		t.Fatalf("Could not read test file: %v", err)
	}
	data = data[44:]

	var transcript string

	frameLengthInBytes := FrameLength * 2
	numFrames := len(data) / frameLengthInBytes
	pcm := make([]int16, FrameLength)
	for i := 0; i < numFrames; i++ {
		for j := 0; j < FrameLength; j++ {
			offset := (i * frameLengthInBytes) + (j * 2)
			pcm[j] = int16(binary.LittleEndian.Uint16(data[offset : offset+2]))
		}

		partial, _, err := cheetah.Process(pcm)
		if err != nil {
			t.Fatalf("Failed to process pcm buffer: %v", err)
		}

		transcript += partial
	}

	final, err := cheetah.Flush()
	if err != nil {
		t.Fatalf("Failed to flush: %v", err)
	}
	transcript += final

	var normalizedTranscript = referenceTranscript
	if !enableAutomaticPunctuation {
		for _, punctuation := range punctuations {
			normalizedTranscript = strings.ReplaceAll(normalizedTranscript, punctuation, "")
		}
	}

	errorRate := getWordErrorRate(transcript, normalizedTranscript)
	if errorRate >= targetErrorRate {
		t.Fatalf("Expected '%f' got '%f'", targetErrorRate, errorRate)
	}
}

func TestProcess(t *testing.T) {
	for _, test := range languageTests {
		runProcessTestCase(t, test.language, test.testAudioFile, test.transcript, test.punctuations, test.errorRate, false)
	}
}

func TestProcessWithPunctuation(t *testing.T) {
	for _, test := range languageTests {
		runProcessTestCase(t, test.language, test.testAudioFile, test.transcript, test.punctuations, test.errorRate, true)
	}
}

func TestMessageStack(t *testing.T) {
	cheetah = NewCheetah("invalid")
	cheetah.EnableAutomaticPunctuation = true
	err := cheetah.Init()
	if err == nil {
		log.Fatalf("Expected Cheetah init failure")
	}

	err = cheetah.Init()
	err2 := cheetah.Init()

	if len(err.Error()) > 1024 {
		t.Fatalf("length of error is full: '%d'", len(err.Error()))
	}

	if len(err2.Error()) != len(err.Error()) {
		t.Fatalf("length of 1st init '%d' does not match 2nd init '%d'", len(err.Error()), len(err2.Error()))
	}
}

func TestProcessFlushMessageStack(t *testing.T) {
	cheetah = NewCheetah(testAccessKey)
	cheetah.EnableAutomaticPunctuation = true
	err := cheetah.Init()
	if err != nil {
		log.Fatalf("Failed to init cheetah with: %v", err)
	}

	address := cheetah.handle
	cheetah.handle = nil

	testPcm := make([]int16, FrameLength)

	_, _, err = cheetah.Process(testPcm)
	if err == nil {
		t.Fatalf("Expected cheetah process to fail")
	}

	_, err = cheetah.Flush()
	if err == nil {
		t.Fatalf("Expected cheetah flush to fail")
	}

	cheetah.handle = address
	delErr := cheetah.Delete()
	if delErr != nil {
		t.Fatalf("%v", delErr)
	}
}

File: binding/java/test/ai/picovoice/cheetah/CheetahTest.java
/*
    Copyright 2022-2025 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.cheetah;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.junit.jupiter.api.Test;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.stream.Stream;


import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class CheetahTest {
    private final String accessKey = System.getProperty("pvTestingAccessKey");
    private final String device = System.getProperty("pvTestingDevice");

    private static int levenshteinDistance(String[] transcript, String[] reference) {
        int m = transcript.length;
        int n = reference.length;
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 0; i <= m; i++) {
            dp[i][0] = i;
        }

        for (int j = 0; j <= n; j++) {
            dp[0][j] = j;
        }

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (transcript[i - 1].equalsIgnoreCase(reference[j - 1])) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = 1 + Math.min(dp[i - 1][j - 1], Math.min(dp[i - 1][j], dp[i][j - 1]));
                }
            }
        }
        return dp[m][n];
    }

    private static float getErrorRate(String transcript, String reference) {
        String[] transcriptWords = transcript.split("\\s+");
        String[] referenceWords = reference.split("\\s+");
        int distance = levenshteinDistance(transcriptWords, referenceWords);

        return (float) distance / (float) referenceWords.length;
    }

    private static ProcessTestData[] loadProcessTestData() throws IOException {
        final Path testDataPath = Paths.get(System.getProperty("user.dir"))
                .resolve("../../resources/.test")
                .resolve("test_data.json");
        final String testDataContent = new String(Files.readAllBytes(testDataPath), StandardCharsets.UTF_8);
        final JsonObject testDataJson = JsonParser.parseString(testDataContent).getAsJsonObject();

        final JsonArray testParameters = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("language_tests");

        final ProcessTestData[] processTestData = new ProcessTestData[testParameters.size()];
        for (int i = 0; i < testParameters.size(); i++) {
            final JsonObject testData = testParameters.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String testAudioFile = testData.get("audio_file").getAsString();
            final String transcript = testData.get("transcript").getAsString();
            final float errorRate = testData.get("error_rate").getAsFloat();

            final JsonArray punctuationsJson = testData.getAsJsonArray("punctuations");
            final String[] punctuations = new String[punctuationsJson.size()];
            for (int j = 0; j < punctuationsJson.size(); j++) {
                punctuations[j] = punctuationsJson.get(j).getAsString();
            }

            final JsonArray modelsJson = testData.getAsJsonArray("models");
            final String[] models = new String[modelsJson.size()];
            for (int j = 0; j < modelsJson.size(); j++) {
                models[j] = modelsJson.get(j).getAsString();
            }

            processTestData[i] = new ProcessTestData(
                    language,
                    models,
                    testAudioFile,
                    transcript,
                    punctuations,
                    errorRate);
        }
        return processTestData;
    }

    private static Stream<Arguments> processTestProvider() throws IOException {
        final ProcessTestData[] processTestData = loadProcessTestData();
        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (ProcessTestData processTestDataItem : processTestData) {
            for (String modelPath : processTestDataItem.models) {
                testArgs.add(Arguments.of(
                        processTestDataItem.language,
                        modelPath,
                        processTestDataItem.audioFile,
                        processTestDataItem.transcript,
                        processTestDataItem.punctuations,
                        false,
                        processTestDataItem.errorRate));
                testArgs.add(Arguments.of(
                        processTestDataItem.language,
                        modelPath,
                        processTestDataItem.audioFile,
                        processTestDataItem.transcript,
                        processTestDataItem.punctuations,
                        true,
                        processTestDataItem.errorRate));
            }
        }

        return testArgs.stream();
    }

    @Test
    void getVersion() throws CheetahException {
        Cheetah cheetah = new Cheetah.Builder()
                .setAccessKey(accessKey)
                .setDevice(device)
                .build();
        assertTrue(cheetah.getVersion() != null && !cheetah.getVersion().equals(""));
        cheetah.delete();
    }

    @Test
    void getFrameLength() throws CheetahException {
        Cheetah cheetah = new Cheetah.Builder()
                .setAccessKey(accessKey)
                .setDevice(device)
                .build();
        assertTrue(cheetah.getFrameLength() > 0);
        cheetah.delete();
    }

    @Test
    void getSampleRate() throws CheetahException {
        Cheetah cheetah = new Cheetah.Builder()
                .setAccessKey(accessKey)
                .setDevice(device)
                .build();
        assertTrue(cheetah.getSampleRate() > 0);
        cheetah.delete();
    }

    @Test
    void getErrorStack() {
        String[] error = {};
        try {
            new Cheetah.Builder()
                    .setAccessKey("invalid")
                    .setDevice(device)
                    .build();
        } catch (CheetahException e) {
            error = e.getMessageStack();
        }

        assertTrue(0 < error.length);
        assertTrue(error.length <= 8);

        try {
            new Cheetah.Builder()
                    .setAccessKey("invalid")
                    .setDevice(device)
                    .build();
        } catch (CheetahException e) {
            for (int i = 0; i < error.length; i++) {
                assertEquals(e.getMessageStack()[i], error[i]);
            }
        }
    }

    @ParameterizedTest(name = "test process data for ''{1}'' with punctuation ''{5}''")
    @MethodSource("processTestProvider")
    void process(
            String language,
            String modelFile,
            String testAudioFile,
            String referenceTranscript,
            String[] punctuations,
            boolean enableAutomaticPunctuation,
            float targetErrorRate) throws Exception {
        String modelPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../lib/common/%s", modelFile))
                .toString();

        Cheetah cheetah = new Cheetah.Builder()
                .setAccessKey(accessKey)
                .setModelPath(modelPath)
                .setDevice(device)
                .setEnableAutomaticPunctuation(enableAutomaticPunctuation)
                .build();

        int frameLen = cheetah.getFrameLength();
        String audioFilePath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../resources/audio_samples/%s", testAudioFile))
                .toString();
        File testAudioPath = new File(audioFilePath);

        AudioInputStream audioInputStream = AudioSystem.getAudioInputStream(testAudioPath);
        assertEquals(16000, audioInputStream.getFormat().getFrameRate());

        int byteDepth = audioInputStream.getFormat().getFrameSize();
        byte[] pcm = new byte[frameLen * byteDepth];
        short[] cheetahFrame = new short[frameLen];

        StringBuilder transcript = new StringBuilder();
        int numBytesRead = 0;
        while ((numBytesRead = audioInputStream.read(pcm)) != -1) {
            if (numBytesRead / byteDepth == frameLen) {
                ByteBuffer.wrap(pcm).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(cheetahFrame);
                CheetahTranscript transcriptObj = cheetah.process(cheetahFrame);
                transcript.append(transcriptObj.getTranscript());
            }
        }
        CheetahTranscript finalTranscriptObj = cheetah.flush();
        transcript.append(finalTranscriptObj.getTranscript());

        cheetah.delete();

        String normalizedTranscript = referenceTranscript;
        if (!enableAutomaticPunctuation) {
            for (String punctuation : punctuations) {
                normalizedTranscript = normalizedTranscript.replace(punctuation, "");
            }
        }

        assertTrue(getErrorRate(transcript.toString(), normalizedTranscript) < targetErrorRate);
    }

    private static class ProcessTestData {
        public final String language;
        public final String[] models;
        public final String audioFile;
        public final String transcript;
        public final String[] punctuations;
        public final float errorRate;

        public ProcessTestData(
                String language,
                String[] models,
                String audioFile,
                String transcript,
                String[] punctuations,
                float errorRate) {
            this.language = language;
            this.models = models;
            this.audioFile = audioFile;
            this.transcript = transcript;
            this.punctuations = punctuations;
            this.errorRate = errorRate;
        }
    }
}
